apm-agent-core/src/test/java/co/elastic/apm/agent/impl/context/ServiceTargetTest.java

```diff
@@ -0,0 +1,217 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.impl.context;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static co.elastic.apm.agent.testutils.assertions.Assertions.assertThat;
+
+class ServiceTargetTest {
+
+    @Test
+    void createEmpty() {
+        ServiceTarget serviceTarget = new ServiceTarget();
+        assertThat(serviceTarget).isEmpty();
+        assertThat(serviceTarget).isNotSetByUser();
+        assertThat(serviceTarget).hasNoName();
+        assertThat(serviceTarget.getType()).isNull();
+    }
+
+    @Test
+    void typeOnly() {
+        ServiceTarget serviceTarget = new ServiceTarget();
+        String serviceType = "service-type";
+
+        assertThat(serviceTarget.withType(serviceType))
+            .hasType(serviceType)
+            .hasNoName()
+            .hasDestinationResource(serviceType);
+
+        assertThat(serviceTarget).isNotSetByUser();
+
+        assertThat(serviceTarget.hasContent()).isTrue();
+
+        serviceTarget.resetState();
+        assertThat(serviceTarget).isEmpty();
+    }
+
+    @Test
+    void typeAndName() {
+        ServiceTarget serviceTarget = new ServiceTarget();
+
+        assertThat(serviceTarget.withType("service-type").withName("service-name"))
+            .hasType("service-type")
+            .hasName("service-name");
+
+        assertThat(serviceTarget)
+            .hasDestinationResource("service-type/service-name");
+
+        assertThat(serviceTarget.getDestinationResource())
+            .describedAs("returned value should be cached without modifications")
+            .isSameAs(serviceTarget.getDestinationResource());
+
+        assertThat(serviceTarget)
+            .isNotSetByUser();
+
+        assertThat(serviceTarget.hasContent()).isTrue();
+
+        serviceTarget.resetState();
+        assertThat(serviceTarget).isEmpty();
+    }
+
+    @Test
+    void userValuePriority() {
+
+        ServiceTarget serviceTarget = new ServiceTarget();
+        serviceTarget.withType("type").withName("name");
+        assertThat(serviceTarget).hasType("type").hasName("name").isNotSetByUser();
+
+        assertThat(serviceTarget.withUserType("user-type")).hasType("user-type").isSetByUser();
+        assertThat(serviceTarget.withUserName("user-name")).hasName("user-name").isSetByUser();
+    }
```

Comment:
Test in reverse order as well (user value first) and test `withHostPortName` as well